teps/0004-tekton-integrations-annotations-statuses.md

```diff
@@ -0,0 +1,689 @@
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy.  Anything
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).
+-->
+
+Integrations will be responsible for handling testing of their own components.
+For pipelines, this is mainly the storage of additional metadata. As long as we
+verify that it is preserved through a pipeline that is sufficient.
+
+## Drawbacks
+
+<!--
+Why should this TEP _not_ be implemented?
+-->
+
+### Tekton Chains
+
+```
+<<[UNRESOLVED dlorenc ]>>
+This section needs more clarification. Depending on how Tekton Chains operates, this might be okay.
+<<[/UNRESOLVED]>>
+```
+
+Tekton Chains currently calls out storing TaskRun definition data for verifiable
+builds. It is unclear if this includes the entire TaskRun definition, or just
+the `TaskRunSpec`.
+
+If it includes `TaskRunStatus` data this proposal will be at odds with Tekton
+Chains, since fundamentally this processes and annotates integration status
+post-pipeline execution and is not guaranteed to complete before chains
+snapshots and signs TaskRun data.
+
+If `TaskRunStatus` is omitted, then there are no issues.
+
+### Reconcile loops
+
+If integrators are not careful, they may accidentally trigger reconcile loops if
+they are constantly updating state based on resource updates (e.g. notice
+update, record new Generation version, update metadata with the version causing
+a new Generation). This is a risk for any pattern that is self-updating a
+resource, so no particular recommendations here beyond "don't do this", but it
+is something I want to call out in case there are known best practices here.
+
+## Alternatives
+
+<!--
+What other approaches did you consider and why did you rule them out?  These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+### Automatically injecting Tasks
+
+This alternative keeps the idea of running user tasks inside of the Pipeline,
+but we make it easier for users by automatically appending integration tasks to
+the end of their pipeline / adding them as finally tasks.
+
+This has a few drawbacks:
+
+- We can't call tasks if the pipeline never runs.
+- The user pipeline needs to have access to all secrets, which limits what we
+  can do with shared multitenant secrets.
+
+### Storing integration statuses outside of Run status
+
+As an alternative to storing integration status information within a
+Task/PipelineRun status, we could store this information in another CRD type
+that references back to the run. The advantage of this is that it avoids making
+an API change to the Task/PipelineRun object directly and the consequences that
+may arise.
```

Comment:
I don't think we need another type for this. It would be on the *integration* itself to decide how it wants to store/display/… the status.